### ISSUE TITLE
Quoted next to / vs Binary String

### ISSUE THREAD

**Issue by ktomk** [sebastianbergmann/exporter#34]
*request for comments*

I make use of the exporter library via Phpunit.

When I'm looking for expected output and this fails and the output contains characters from the C0 / C1 range (control characters), the string is hex-dumped.

This is kinda cumbersome as it is not easily readable any longer. I can copy the hex-string (after beginning 0x) and convert it to a readable string then, e.g.:

~~~
$ xclip -o -selection cliboard | xxd -r -p | less
~~~

This works but needs extra manual labour.

In these cases I can also patch exporter by replacing in `src/Exporter.php`

~~~
                return 'Binary String: 0x' . \bin2hex($value);
~~~

with

~~~
                return 'Quoted String: "' . \addcslashes($value, "\0..\11!@\14..\37!@\177..\377") . '"';
~~~

which perfectly fit my needs. This is especially well fitting in my case as I have merely standard output with very little control characters (mainly flow control) of which some are fine to render (like \n) and some are fine to transpose (with the example code above in octal notation).

I'm filing this issue as I'd like to get some comments on this. My main aim here is readability by the user which I think the exporter is for but this is a broad topic and needs may vary across users.

Also I'm asking myself what a good strategy would be. For pure binary strings the octal representation might be more verbose than the hexits so the hex notation is still preferable for purely binary strings.

The routine could look for a NUL byte in the string and if that is the case have it like before as "Binary String: 0x....". This can be compared on how grep does decide whether a file is binary or not.

I normally only need from Unicode the part it shares in the UTF-8 variant with the US-ASCII character set, so UTF-8 encoding is not specifically addressed here but it might play a role for others, too.

**Comment by sebastianbergmann** [comment]
@ktomk Can you test whether #91 solves this for you? Thank you.

**Comment by ktomk** [comment]
> [@ktomk](https://github.com/ktomk) Can you test whether [#91](https://github.com/sebastianbergmann/exporter/pull/91) solves this for you? Thank you.

I took a look and so far it looks good to me, have to remember what the test-cases were for solid testing. The only thing that strove my eyes was `"\n\r" => '\n\r' . "\n",` in the translation table because `"\n\r"` normally is not a thing, cases are \r\n (win), \r (old mac), \n (unix), with the consequence that eventually lone-standing \n followed by \r are terminated with a single \n instead of both individually.

**Comment by sebastianbergmann** [comment]
Good catch, I cleaned that up in b54c4ddf234bbaa3de81fa69461c5830147c7a12.
